README.md

```diff
@@ -19,191 +14,233 @@ pip install promote
 
-# Send a request to the model
-p.predict("HelloModel", {"name": "greg"})
+**Arguments**
+- `data`(_list_ or _dict_): the parsed JSON sent to the deployed model
 
-# { 'result': {'response': 'Hello greg!'}},
-# 'version': 1,
-# 'promote_id': '9fbcb56a89bf318d8b50c41654554e3c',
-# 'promote_model': 'HelloModel'}
+**Example**
+```python
+modelFunction <- function(data) {
+  # generate predictions from the model based on the incoming data
+  my_model.predict(data)
+}
 ```
-
 <hr>
 
-### The `@promote.validate_json` decoractor
-
-`@promote.validate_json` is an optional decorator that can be useful for validating data as it comes into a model.  It utilizes the `schema` package to validate JSON.
+#### <a name="validate"></a>`@promote.validate_json()`
+It is possible to decorate your model function with the `promote.validate_json` decorator. This validates that the input data to the model meets specific predefined criteria. Failure to meet these criteria will throw an error.
 
-#### Usage
-
-`from schema import Schema`
-`@promote.validate_json(Schema({'name': And(str, len)}))`
+**Arguments**
+- `aSchema`(_Schema_): a valid `schema.Schema` object
 
+**Example**
 ```python
-import promote
+from schema import Schema, And
 
-@promote.validate_json(Schema({'name': And(str, lambda s: len(s) > 1)}))
-def helloModel(data):
-    return {'response': 'Hello ' + data['name'] + '!'}
+@promote.validate_json(Schema({'X1': And(int, lambda s: min([t > 0 for t in s]))},{'X2': And(int, lambda s: min([t > 0 for t in s]))}))
+modelFunction <- function(data) {
+  my_model.predict(data)
+}
 ```
+<hr>
 
-The code above ensures that all data passed to our model must:
-
-1. Be an object, `{}`
-2. Contain the key `'name'`
-3. Contain a string with length greater than 1 character.
-
-### Setting the Auth
+#### <a name="testing"></a>Test Data
+It is a good practice to test the model function as part of the deployment script to make sure it successfully produces an output. Once deployed, the `data` being input into the model function will always be in the form of a python [dict](https://docs.python.org/3/tutorial/datastructures.html#dictionaries) or [list](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists). The incoming JSON string will be parsed using the `loads()` method available from the [json](https://docs.python.org/3/library/json.html) library.
 
-To deploy models, you'll need to add your username, API key, and URL of Promote to a new instance of the class `Promote`
+**Example**
 ```python
-p = promote.Promote(USERNAME, API_KEY, PROMOTE_URL)
+testdata <- '{"X1":[1,2,3],"X2":[4,5,6]}'
+modelFunction(json.loads(testdata))
+
 ```
 <hr>
 
-### `Promote`
+#### `promote.Promote()`
+To deploy models, you'll need to add your username, API key, and URL of Promote to a new instance of the class `Promote`
 
-The `Promote` module has 3 methods:
+**Arguments**
```

Comment:
SImilar to R, do we want to add the full function signature here? I think it would be useful, especially with the parameters with default values it shows them at a glance instead of having to read through the arguments list individually. Also in Python the order is more important than R's optional parameters.

Reply:
Yes let's do it, ensuring that we have the parameter names correct